mat64/io.go

```diff
@@ -81,18 +87,24 @@ func (m *Dense) UnmarshalBinary(data []byte) error {
 		panic("mat64: unmarshal into non-zero matrix")
 	}
 
+	var tmp [8]byte
 	buf := bytes.NewReader(data)
-	var rows int64
-	err := binary.Read(buf, binary.LittleEndian, &rows)
+	_, err := buf.Read(tmp[:])
```

Comment:
I tried with:

```go
// readFull reads exactly len(buf) bytes from r into buf.
// It returns an error if fewer bytes were read.
// If an EOF happens after reading some but not all the bytes, readFull
// returns io.ErrUnexpectedEOF
//
// readFull is heavily inspired from io.ReadFull.
func readFull(r io.Reader, buf []byte) error {
        n, err := r.Read(buf)
        if n != len(buf) {
               return errors.New("mat64: short read")
        }
        if err == io.EOF {
                err = io.ErrUnexpectedEOF
        }
        return err
}
```

this gave me (wrt sans-ReadFull):

```
name                    old time/op    new time/op    delta
UnmarshalDense10-4         425ns ± 1%     711ns ± 2%   +67.47%          (p=0.008 n=5+5)
UnmarshalDense100-4       3.11µs ± 1%    4.29µs ± 1%   +37.89%          (p=0.008 n=5+5)
UnmarshalDense1000-4      29.2µs ± 1%    40.1µs ± 2%   +37.64%          (p=0.008 n=5+5)
UnmarshalDense10000-4      281µs ± 2%     400µs ± 1%   +42.49%          (p=0.008 n=5+5)
UnmarshalVector10-4        411ns ± 1%     642ns ± 0%   +56.27%          (p=0.016 n=5+4)
UnmarshalVector100-4      3.13µs ± 4%    4.22µs ± 4%   +34.95%          (p=0.008 n=5+5)
UnmarshalVector1000-4     29.6µs ± 3%    39.1µs ± 3%   +32.19%          (p=0.008 n=5+5)
UnmarshalVector10000-4     290µs ± 2%     373µs ± 0%   +28.49%          (p=0.016 n=5+4)

name                    old alloc/op   new alloc/op   delta
UnmarshalDense10-4         80.0B ± 0%    136.0B ± 0%   +70.00%          (p=0.008 n=5+5)
UnmarshalDense100-4         896B ± 0%      952B ± 0%    +6.25%          (p=0.008 n=5+5)
UnmarshalDense1000-4      8.19kB ± 0%    8.25kB ± 0%    +0.68%          (p=0.008 n=5+5)
UnmarshalDense10000-4     81.9kB ± 0%    82.0kB ± 0%    +0.07%          (p=0.016 n=4+5)
UnmarshalVector10-4        80.0B ± 0%    136.0B ± 0%   +70.00%          (p=0.008 n=5+5)
UnmarshalVector100-4        896B ± 0%      952B ± 0%    +6.25%          (p=0.008 n=5+5)
UnmarshalVector1000-4     8.19kB ± 0%    8.25kB ± 0%    +0.68%          (p=0.008 n=5+5)
UnmarshalVector10000-4    81.9kB ± 0%    82.0kB ± 0%    +0.07%          (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
UnmarshalDense10-4          1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense100-4         1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense1000-4        1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense10000-4       1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector10-4         1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector100-4        1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector1000-4       1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector10000-4      1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
```

then, with `readFull`:

```go
func readFull(r io.Reader, buf []byte) error {
        _, err := r.Read(buf)
        if err == io.EOF {
                err = io.ErrUnexpectedEOF
        }
        return err
}
```

and got:
```
name                    old time/op    new time/op    delta
UnmarshalDense10-4         425ns ± 1%     683ns ± 1%   +60.83%          (p=0.008 n=5+5)
UnmarshalDense100-4       3.11µs ± 1%    4.11µs ± 1%   +31.88%          (p=0.008 n=5+5)
UnmarshalDense1000-4      29.2µs ± 1%    37.6µs ± 1%   +29.05%          (p=0.008 n=5+5)
UnmarshalDense10000-4      281µs ± 2%     366µs ± 1%   +30.13%          (p=0.008 n=5+5)
UnmarshalVector10-4        411ns ± 1%     638ns ± 1%   +55.18%          (p=0.008 n=5+5)
UnmarshalVector100-4      3.13µs ± 4%    4.10µs ± 2%   +31.33%          (p=0.008 n=5+5)
UnmarshalVector1000-4     29.6µs ± 3%    37.9µs ± 1%   +27.97%          (p=0.008 n=5+5)
UnmarshalVector10000-4     290µs ± 2%     368µs ± 2%   +26.85%          (p=0.008 n=5+5)

name                    old alloc/op   new alloc/op   delta
UnmarshalDense10-4         80.0B ± 0%    136.0B ± 0%   +70.00%          (p=0.008 n=5+5)
UnmarshalDense100-4         896B ± 0%      952B ± 0%    +6.25%          (p=0.008 n=5+5)
UnmarshalDense1000-4      8.19kB ± 0%    8.25kB ± 0%    +0.68%          (p=0.008 n=5+5)
UnmarshalDense10000-4     81.9kB ± 0%    82.0kB ± 0%    +0.07%          (p=0.016 n=4+5)
UnmarshalVector10-4        80.0B ± 0%    136.0B ± 0%   +70.00%          (p=0.008 n=5+5)
UnmarshalVector100-4        896B ± 0%      952B ± 0%    +6.25%          (p=0.008 n=5+5)
UnmarshalVector1000-4     8.19kB ± 0%    8.25kB ± 0%    +0.68%          (p=0.008 n=5+5)
UnmarshalVector10000-4    81.9kB ± 0%    82.0kB ± 0%    +0.07%          (p=0.008 n=5+5)

name                    old allocs/op  new allocs/op  delta
UnmarshalDense10-4          1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense100-4         1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense1000-4        1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalDense10000-4       1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector10-4         1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector100-4        1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector1000-4       1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
UnmarshalVector10000-4      1.00 ± 0%      3.00 ± 0%  +200.00%          (p=0.008 n=5+5)
```